noir-projects/aztec-nr/aztec/src/state_vars/shared_mutable/scheduled_delay_change.nr

```diff
@@ -0,0 +1,487 @@
+            // that overall the current delay is respected.
+            //
+            //      current              delay              earliest value block of change
+            //       block           block of change         if delay remained unchanged
+            //  =======N===================|================================X=================>
+            //         ^                   ^                                ^
+            //         |-------------------|--------------------------------|
+            //         |  schedule delay                new delay           |
+            //         ------------------------------------------------------
+            //                            current delay
+            current - new
+        };
+
+        self.pre = Option::some(current);
+        self.post = Option::some(new);
+        self.block_of_change = current_block_number + schedule_delay;
+    }
+
+    /// Returns the minimum delay before a value might mutate due to a scheduled change, from the perspective of some
+    /// historical block number. It only returns a meaningful value when called in private with historical blocks. This 
+    /// function can be used alongside `ScheduledValuechange.get_block_horizon` to properly constrain the
+    /// `max_block_number` transaction property when reading mutable shared state.
+    /// This value typically equals the current delay at the block following the historical one (the earliest one in
+    /// which a value change could be scheduled), but it also considers scenarios in which a delay reduction is 
+    /// scheduled to happen in the near future, resulting in a way to schedule a change with an overall delay lower than
+    /// the current one.
+    pub fn get_effective_minimum_delay_at(self, historical_block_number: u32) -> u32 {
+        if self.block_of_change <= historical_block_number {
+            // If no delay changes were scheduled, then the delay value at the historical block (post) is guaranteed to
+            // hold due to how further delay changes would be scheduled by `schedule_change`.
+            self.post.unwrap_or(INITIAL_DELAY)
+        } else {
+            // If a change is scheduled, then the effective delay might be lower than the current one (pre). At the
+            // block of change the current delay will be the scheduled one, with an overall delay from the historical
+            // block number equal to the number of blocks until the change plus the new delay. If this value is lower
+            // than the current delay, then that is the effective minimum delay.
+            //
+            //
+            //     historical       delay                 actual earliest value
+            //       block      block of change              block of change
+            //  =======NS=============|=============================X===========Y=====>
```

Comment:
`N` is the historical block number, and `S` is the block when the schedule happens. It's annoying because both blocks are important, and they are right next to each other. I added an arrow to try to clarify this